### DISCUSSION TITLE
Installing Python bindings

### DISCUSSION THREAD

**Discussion by nilsvu** [spack/spack#28258] [Q&A] [answered]
Is there an established way to install Python bindings of a C++ program with Spack? Can someone point out existing documentation, or examples of Spack packages that do this well? Essentially, a program I'm writing a Spack package for has a `BUILD_PYTHON_BINDINGS` CMake toggle that produces a directory like this:

```
python/
  setup.py
  my_package/
    foo.py
    bar.py
```

If I built the program manually, I would now create a `python -m venv` and `pip install ./python/`. Since `setup.py` lists a bunch of Python dependencies, those would install alongside `my_package` automatically and the package is ready for use. With Spack, it would be great to achieve a similar result.

**Comment by nilsvu** [comment, marked as answer]
My understanding from reading packages like `vtk` is now that they make sure to install their bindings in

```
prefix/lib/pythonX.Y/site-packages/
```

and then they

```
extend('python')
```

so their prefix gets symlinked into Python. Furthermore, they don't use the information from `setup.py` at all but list all Python dependencies in the Spack package.

**Reply by adamjstewart** [reply]
To add to this, you can also use pip to install things. For example, if the build phase creates the directory structure you listed above, you can add:
```python
depends_on('py-pip', type='build')

@run_after('install')
def install_python(self):
    args = std_pip_args + ['--prefix=' + self.prefix, 'python']
    pip(*args)
```

> Since `setup.py` lists a bunch of Python dependencies, those would install alongside `my_package` automatically and the package is ready for use.
>
> Furthermore, they don't use the information from `setup.py` at all but list all Python dependencies in the Spack package.

This is partially true. We (as humans) use `setup.py` to find the list of dependencies to add to Spack's `package.py`, so that Spack (as a package manager) will install them for us. We don't want to rely on `pip` to find and install dependencies for us as 1) this won't work on air-gapped systems without internet access, and 2) it may install conflicting versions of a dependency compared to what is already in the DAG, and 3) it isn't reproducible in any way.

P.S. I answered a similar question in #30757, but if you have any other Python-related questions let me know!
